web-api/src/storage/memory.rs

```diff
@@ -83,26 +88,69 @@ impl TryFrom<&DocumentId> for models::DocumentId {
 
 #[derive(Clone, Debug, Deserialize, Serialize)]
 struct Document {
-    embedding: Embedding,
     properties: DocumentProperties,
     category: Option<String>,
 }
 
-impl From<(models::DocumentId, &Document)> for PersonalizedDocument {
-    fn from((document_id, document): (models::DocumentId, &Document)) -> Self {
-        Self {
-            id: document_id,
-            score: 0.,
-            embedding: document.embedding.clone(),
-            properties: document.properties.clone(),
-            category: document.category.clone(),
+#[derive(Clone, Copy, Debug, Deref)]
+struct EmbeddingRef<'a>(&'a Embedding);
+
+impl Point for EmbeddingRef<'_> {
+    fn distance(&self, other: &Self) -> f32 {
+        1. - cosine_similarity(self.view(), other.view())
+    }
+}
+
+#[self_referencing]
+struct Embeddings {
+    map: FnvHashMap<DocumentId, Embedding>,
+    #[borrows(map)]
+    #[covariant]
+    index: HnswMap<EmbeddingRef<'this>, &'this DocumentId>,
+}
+
+impl From<FnvHashMap<DocumentId, Embedding>> for Embeddings {
```

Comment:
:art: ? Given how computation intensive this can be maybe it's not the best choice to place it in a `From` impl. it kinda makes it easy to overlook that `document.1 = embeddings.into();` isn't just some trivial conversion.